labs/cryptol-overview/lab-xx-overview.md

```diff
@@ -0,0 +1,349 @@
+
+We can see the effect of encrypting the particular input `attack at dawn` with the key `0xff`:
+
+```haskell
+Main> encrypt 0xff "attack at dawn"
+[0x9e, 0x8b, 0x8b, 0x9e, 0x9c, 0x94, 0xdf, 0x9e, 0x8b, 0xdf, 0x9b,
+ 0x9e, 0x88, 0x91]
+```
+
+Cryptol interprets the string `"attack at dawn"` as a sequence of bytes suitable for the encrypt operations. We will introduce Cryptol types in this lab and discuss them in detail throughout this course.
+
+Furthermore, we can prove this property holds in the interpreter using the `:prove` command and the currently configured SAT solver (Z3 by default):
+
+```haskell
+Main> :prove roundtrip : [8] -> [16][8] -> Bit
+Q.E.D.
+(Total Elapsed Time: 0.010s, using Z3)
+```
+
+Cryptol reports `Q.E.D.` indicating that our property is indeed true for all possible inputs. Here we must specify the length of the messages that we want to check this property for. This example checks the property for 16 character messages, but we could check this for any arbitrary length.
+
+
+# Language Features
+
+So what makes Cryptol special compared to other langauges?
+
+Cryptol is a language designed with Cryptography specifically in mind -- much of the syntax and language was designed with the way that real cryptographers think about and design systems. This allows the Cryptol user to create formal algorithm specifications that closely imitate the style used to describe these algorithms mathematically.
+
+Furthermore, Cryptol provides direct access and easily integrates with powerful tools such as SAT solvers and the Software Analysis Workbench (SAW). These tools allow the user to *prove* facts and demonstrate properties about their code which can provide assurance guarantees that go far beyond simple unit testing.
+
+We will introduce some of these features below and discuss how they support building Cryptographic specifications and evaluations. If you have access to the Cryptol interpreter you can follow along with some of the examples, but a detailed introduction to the Cryptol interpreter will be introduced in future lessons.
+
+
+## Basic Data Types
+
+Cryptol was designed to provide easy access to the sorts of data and operations that appear in Cryptographic algorithms and specifications. There are five basic data types provided by Cryptol: bits, sequences, integers, tuples, and records. Cryptol also supports the ability to create user defined types built up from the basic types. In this section we present some basic examples deomonstrating these types, note that commands using `:t` are a request to Cryptol to report the type of argument.
+
+ * **Bits** - The simplest data type which can take on two values, `True` and `False`. Common operations like `and`, `or`, and `not` are available.
+
+```haskell
+Cryptol> :t True
+True : Bit
+Cryptol> True && False
+False
+Cryptol> True /\ False
+False
+Cryptol> True || False
+True
+Cryptol> True \/ False
+True
+Cryptol> ~True
+False
```

Comment:
The way this is presented may confuse people into thinking that `&&` and `/\` are synonymous (same for `||` and `\/`).